File: solutions/162.Find_Peak_Element/AC_binary_search.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# Author:      illuz <iilluzen[at]gmail.com>
# File:        AC_binary_search.py
# Create Date: 2015-02-14 09:42:56
# Usage:       AC_binary_search.py 
# Descripton:  


class Solution:
    def findPeakInRange(self, num, lhs, rhs):
        mid = (lhs + rhs) // 2
        if (mid == lhs or num[mid] > num[mid - 1]) and \
            (mid == rhs or num[mid] > num[mid + 1]):
            return mid
        if num[mid] < num[mid + 1]:
            return self.findPeakInRange(num, mid + 1, rhs)
        else:
            return self.findPeakInRange(num, lhs, mid)
    # @param num, a list of integer
    # @return an integer
    def findPeakElement(self, num):
        return self.findPeakInRange(num, 0, len(num) - 1)


# debug
s = Solution()
print s.findPeakElement([1, 2, 3, 1])

File: solutions/081.Search_in_Rotated_Sorted_Array_II/AC_binary_search.cpp
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_binary_search.cpp
*  Create Date: 2015-01-27 10:42:49
*  Descripton:  Like the I version
*/

#include <bits/stdc++.h>

using namespace std;
const int N = 0;

class Solution {
public:
    bool search(int A[], int n, int target) {
        int left = 0, right = n;
        while (left != right) {
            int mid = (left + right) / 2;
            if (A[mid] == target)
                return true;
            if (A[mid] > A[left]) {
                // [left~mid] is unrotated sorted
                if (A[left] <= target && target < A[mid])
                    right = mid;
                else
                    left = mid + 1;
            } else if (A[mid] < A[left]) {
                // [mid~right] is unrotated sorted
                if (A[mid] < target && target <= A[right - 1])
                    left = mid + 1;
                else
                    right = mid;
            } else    // skip
                left++;
        }
        return false;
    }
};

int main() {
    int n, A[100], tar;
    Solution s;
    while (cin >> n) {
        for (int i = 0; i < n; i++)
            cin >> A[i];
        cin >> tar;
        cout << s.search(A, n, tar) << endl;
    }
    return 0;
}

